libs/src/sdk/api/tracks/TracksApi.ts

```diff
@@ -0,0 +1,128 @@
+import { BASE_PATH, RequiredError } from '../generated/default/runtime'
+
+import {
+  Configuration,
+  StreamTrackRequest,
+  TracksApi as GeneratedTracksApi
+} from '../generated/default'
+import type { DiscoveryNodeSelectorService } from '../../services/DiscoveryNodeSelector'
+import type { UploadTrackRequest } from './types'
+import type { StorageService } from '../../services/StorageService'
+import { isFileValid } from '../../utils/file'
+import { TRACK_REQUIRED_PROPERTIES, TRACK_REQUIRED_VALUES } from './constants'
+import { objectMissingProperties } from '../../utils/object'
+
+// Subclass type masking adapted from Damir Arh's method:
+// https://www.damirscorner.com/blog/posts/20190712-ChangeMethodSignatureInTypescriptSubclass.html
+// Get the type of the generated TracksApi excluding streamTrack
+type GeneratedTracksApiWithoutStream = new (config: Configuration) => {
+  [P in Exclude<keyof GeneratedTracksApi, 'streamTrack'>]: GeneratedTracksApi[P]
+}
+
+// Create a new "class" that masks our generated TracksApi with the new type
```

Comment:
what's the reasoning for doing this masking instead of composition - like having TracksApi be its own class that takes a parameter like "discoveryTracksApi". to me this masking seems confusing because it's supposed to be an auto-generated Discovery API, but now it does a bunch of other stuff as well

Reply:
Not 100% sure what the benefits are to doing it this way, this was a change @rickyrombo made and it existed before this pr. It got included in the diff because I moved the file

Reply:
Would love a more composition-based approach, maybe can get your thoughts on it @theoilie 

Right now these are the requirements:
- The autogenerated `/stream` API is pretty garbage, and we want to replace it with our own implementation that basically just returns the route
- We want everything tracks related to be under `sdk.tracks` eg `sdk.tracks.uploadTrack()` and `sdk.tracks.getTrack()`
- We want adding read APIs to the SDK to be as painless as possible, hopefully as easy as running the autogenerator, and always in sync with the actual endpoints

Let's say we use a read API sub-module, `discoveryTracksApi`, and compose `TracksApi` with it and a `storageTracksApi`. Then we make `sdk.tracks` an instance of `TracksApi`. In order to get `sdk.tracks.uploadTrack()` and `sdk.tracks.getTrack()` we would have to re-wrap the `uploadTrack()` method of `storageTracksApi` and `getTrack()` method of discoveryTracksApi` respectfully, passing along their types. Autogeneration wouldn't automatically add new methods to `sdk.tracks` and we could potentially have name mismatches, parameter mismatches, etc all over again.

Reply:
ah I see, thanks for all the context @rickyrombo 

with those requirements, particularly the third one, it does seem pretty hard to do a composition based approach. sounds like you've thought it through pretty thoroughly and made an approach that's pretty minimal and keeps TypeScript happy! it actually looks pretty good - not too complicated!

Reply:
we could probably still compose the write stuff? I haven't looked at this PR in depth yet, but will do by EOD